Breaking upgrade docs (0.4) (#1087)
Quansight/qhub: docs/source/admin_guide/breaking-upgrade.md

# Upgrade - Redeployment for Breaking Changes

> For versions that are known to require redeployment when upgrading, the steps for upgrade include:

Reviewer: ```suggestion
For versions that are known to require redeployment when upgrading,  follow the next steps to upgrade QHub:
```

> - Back up user data by following [this guide](./backup.md).

Reviewer: It is recommended to use numbered lists for steps that must be performed in a given order 
```suggestion
1. Back up user data by following [the backup guide](./backup.md) in the documentation.
```

> - Change existing cluster to a different URL (e.g. `qhub-old.mycompany.com`) so it is hidden.

Reviewer: ```suggestion
- Change the existing cluster to a different URL (e.g. `qhub-old.mycompany.com`) so it is hidden.
```

- Run `qhub upgrade` to make recommended modifications to the `qhub-config.yaml` file.
- Deploy a new QHub with the desired version (to your original preferred URL e.g. `qhub.mycompany.com` but a new project_name to avoid resource name clashes).
- Restore user data to the new cluster.

> - Destroy old cluster when happy.

Reviewer: ```suggestion
- Once you have confirmed the backup and upgrade have been successfully completed, you can destroy the old cluster.
```

> Please always check release notes for more details - and in all cases, backup your data before upgrading.

Reviewer: Can we add a link to where the release notes live?

> **The rest of this guide assumes you are upgrading from version v0.3.14 to v0.4.**

Reviewer: Should this be a note instead of a bold styling? we want to make this as visible as possible

You may be deploying QHub based on a local configuration file, or you may be using CI/CD workflows in GitHub or GitLab. Either way, you will need to locate a copy of your `qhub-config.yaml` configuration file to upgrade it (and commit back to your git repo in the CI/CD case).

For CI/CD deployments, you will need to `git clone <repo URL>` into a folder on your local machine if you haven't done so already.

## 1. Backup existing data

Perform manual backups of the NFS data and JupyterHub database (ignore the section about Keycloak data since that will not exist in your v0.3.14 cluster).

> The [Manual Backup guide is here](./backup.md).

Reviewer: ```suggestion
For more details please refer to the  [Manual Backup guide](./backup.md).
```

## 2. Rename existing QHub URL

This will allow the existing cluster to remain alive in case it is needed, but the idea would be not to have it in use from now on.

In the qhub-config.yaml for example:

> ```

Reviewer: ```suggestion
```yaml
```

project_name: myqhub
domain: qhub.myproj.com
```

could become:

> ```

Reviewer: ```suggestion
```yaml
```

project_name: myqhub
domain: qhub-old.myproj.com
```

> Run qhub deploy using the existing (old) version of qhub.

Reviewer: ```suggestion
Run the command `qhub deploy` using the existing (old) version of QHub.
```

## 3. Upgrade the `qhub` command package

To install (or upgrade) your pip installation of the Python package used to manage QHub:

```shell
pip install --upgrade qhub==0.4.0
```

> ## 4. Upgrade `qhub-config.yaml` file

Reviewer: ```suggestion
## 4. Upgrade the `qhub-config.yaml` file
```

> Now we need to go back to the original `qhub-config.yaml` file which contained the original domain.

Reviewer: There are some inconsistencies between the usage of we and you for instructions

## 5. Rename the Project and Increase Kubernetes version

You need to rename the project to avoid clashes with the existing (old) cluster which would otherwise already own resources based on the names that the new cluster will attempt to use.

The domain should remain as the preferred main one that was always in use previously.

For example:

```
project_name: myqhub
domain: qhub.myproj.com
```

could become:

```
project_name: myqhubnew
domain: qhub.myproj.com
```

> It is also a good time to upgrade your version of Kubernetes. Look for the `kubernetes_version` field within the cloud provider section of the `qhub-config.yaml` file and increase it to the latest.

## 6. Redeploy QHub

You will now have a `qhub-config.yaml` file that you can deploy.

```shell
qhub deploy -c qhub-config.yaml
```

At this point you will see an error message saying that deployment is prevented due to the `prevent_deploy` setting in your YAML file. This is a safeguard to ensure that you only proceed if you are aware of possible breaking changes in the current upgrade.

> Because you are reading this guide, you already know that you needed to backup your data - see above again, if not. So you can edit your `qhub-config.yaml`, find the line reading `prevent_deploy: true` and remove it.

Reviewer: This sentence reads rather unfriendly...

```suggestion
Make sure to **backup your data** as described in the [backup section of the documentation]<add link here>.
Only after backing up your data proceed to remove the `prevent_deploy: true` line in the `qhub-config.yaml` file.
```

Attempt the `qhub deploy -c qhub-config.yaml` command again and it should get further this time.

## 7. CI/CD: render and commit to git

> For CI/CD (GitHub/GitLab) workflows, then as well as generating the updated `qhub-config.yaml` files as above, you will also need to regenerate the workflow files based on the latest `qhub` version's templates.

Reviewer: ```suggestion
For CI/CD (GitHub/GitLab) workflows, as well as generating the updated `qhub-config.yaml` files as above, you will also need to regenerate the workflow files based on the latest `qhub` version's templates.
```

With the newly upgraded `qhub-config.yaml` file, run:

```shell
qhub render -c qhub-config.yaml
```

(Note that `qhub deploy` would have performed this render step too, but will also immediately redeploy your QHub. Run the render command alone if you are now working separately in your repo and don't want to redeploy.)

Commit all the files (`qhub-config.yaml` and GitHub/GitLab workflow files) back to the remote repo. All files need to be committed together in the same commit.

## 8. Update OAuth callback URL for Auth0 or GitHub

If your QHub deployment relies on Auth0 or GitHub for authentication, please update the OAuth callback URL.

> <details><summary>Click for more detailed instructions </summary>

Reviewer: ```suggestion
<details><summary>Expand this section for more detailed instructions </summary>
```

1. Navigate to the your Auth0 tenacy homepage and from there select "Applications".

2. Select the "Regular Web Application" with the name of your deployment.

3. Under the "Application URIs" section, paste the new OAuth callback URL in the "Allowed Callback URLs" text block.

> - The URL will take the shape:

Reviewer: ```suggestion
- The URL will look something like this:
```

    ```
    https://{your-qhub-domain}/auth/realms/qhub/broker/auth0/endpoint
    ```

Updating the GitHub Application instead is very similar, but the OAuth callback will be at `https://{your-qhub-domain}/auth/realms/qhub/broker/github/endpoint`

</details>

## 9. Restore from Backups

> Next, return to the [Manual Backups documentation](./backup.md) to perform:

Reviewer: ```suggestion
Next, you will need to perform the following steps to restore from a previously generated backup as described in the [Manual Backups documentation](./backup.md):
```

# Upgrade - Redeployment for Breaking Changes

For versions that are known to require redeployment when upgrading, this is an outline of the steps required to upgrade QHub:

- Back up user data by following [this backup guide](./backup.md).
- Change the existing cluster to a different URL (for example `qhub-old.mycompany.com`) so it is hidden.
- Run `qhub upgrade` to make recommended modifications to the `qhub-config.yaml` file.
- Deploy a new QHub with the desired version (to your original preferred URL for example `qhub.mycompany.com` but a new project_name to avoid resource name clashes).
- Restore user data to the new cluster.
- Once you have confirmed the backup and upgrade have been successfully completed, you can destroy the old cluster.

Full details of the steps appear below.

> Please always check [release notes](../dev_guide/changelog.md) for more details - and in all cases, backup your data before upgrading.

Reviewer: ```{suggestion}
Please always check the [release notes](../dev_guide/changelog.md) for more details - and in all cases, backup your data before upgrading.
```

> > The rest of this guide assumes you are upgrading from version v0.3.14 to v0.4.

Reviewer: Is this always true or can folks upgrade from any `v0.3.x`? if so it would be best to replace with that

Author: Will add 'or earlier'

Author: Will add 'or earlier'

You may be deploying QHub based on a local configuration file, or you may be using CI/CD workflows in GitHub or GitLab. Either way, you will need to locate a copy of your `qhub-config.yaml` configuration file to upgrade it (and commit back to your git repo in the CI/CD case).

For CI/CD deployments, you will need to `git clone <repo URL>` into a folder on your local machine if you haven't done so already.

## 1. Backup existing data

Perform manual backups of the NFS data and JupyterHub database (ignore the section about Keycloak data since that will not exist in your v0.3.14 cluster).

For more details please refer to the [Manual Backup guide](./backup.md).

## 2. Rename existing QHub URL

This will allow the existing cluster to remain alive in case it is needed, but the idea would be not to have it in use from now on.

> In the qhub-config.yaml for example:

Reviewer: ```{suggestion}
In the `qhub-config.yaml` for example:
```
